### PR TITLE
Intensity augmentation transform

### PR DESCRIPTION
**Pull request by yanielc**
Signed-off-by: yc7620 <yaniel.cabrera20@imperial.ac.uk>

### Description
Intensity augmentation transform to be used as a means of data augmentation for medical imaging datasets. It was first demonstrated in [Intensity Augmentation to Improve Generalizability of Breast Segmentation Across Different MRI Scan Protocols](https://ieeexplore.ieee.org/abstract/document/9166708) that intensity augmentation helps generalize a model. It has been used in other works such as [Image translation for medical image generation: Ischemic stroke lesion segmentation](https://www.sciencedirect.com/science/article/pii/S1746809421008806).

The transform works by replacing the intensities of an image by new values uniformly sampled from all possible intensity values in the input image. Below is an example of the application of the transform on an MRI image:

intensity_remap

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by yanielc** [comment]
Hi, not entirely sure why there are 22 changed files. I only added `transforms/intensity/array.py` and `transforms/__init__.py` to my commit.

File: monai/transforms/intensity/array.py [hunk 1]
```diff
@@ -2053,3 +2055,123 @@ def __call__(self, img: NdarrayOrTensor, mask: Optional[NdarrayOrTensor] = None)
...
+        # smooth
+        noise = torch.nn.AvgPool1d(self.kernel_size, stride=1)(noise.unsqueeze(0)).squeeze()
+        # add linear component
+        grid = torch.arange(len(noise)) / len(noise)
+        noise += self.slope * grid
+        # rescale
+        noise = (noise - noise.min()) / (noise.max() - noise.min()) * img.max() + img.min()
+
+        # intensity remapping function
+        map = dict(zip(vals_to_sample, noise.tolist()))
```
Comment (wyli) [inline monai/transforms/intensity/array.py:2117]: looks good, perhaps implementing this with torch.bucketize https://pytorch.org/docs/stable/generated/torch.bucketize.html could be more efficient...

Comment (yanielc) [reply monai/transforms/intensity/array.py:2117]: Hi @wyli, I looked at the bucketize link you provided, and I don't see how it helps. The line `map = dict(zip(vals_to_sample, noise.tolist()))` is simply creating a lookup table such as { 'old intensity' : 'new intensity' }. Not clear to me how torch.bucketize helps here.

Comment (wyli) [reply monai/transforms/intensity/array.py:2117]: the (untested) idea is that if `vals_to_sample` is sorted monotonically increasing, then `index_img = torch.bucketize(img, vals_to_sample)` will have the indices of every intensity, and then the mapping could be done by `new_img = noise[index_img]`? essentially `for key in map: index_img = (img==key)` is a "one-hot" encoding, `index_img = torch.bucketize(img, vals_to_sample)` is the corresponding non-onehot version. but maybe it doesn't work or it's even slower...

Comment (yanielc) [reply monai/transforms/intensity/array.py]: Using `torch.bucketize` really made the code very efficient. I'll add dictionary versions of the transforms, plus unit tests.

Sorry, I keep seeing that my commit is touching other files...

**Review by wyli — APPROVED** [review]

**Pull request merged**